src/ptvsd/adapter/debuggee.py

```diff
@@ -461,3 +457,45 @@ def _pid_to_handle(pid):
         if not handle:
             raise ctypes.WinError()
         return handle
+
+
+class ProcessPidServer(object):
+ 
+    def __init__(self):
+        self.listener = socket.create_server("127.0.0.1", 0)
+        self.host, self.port = self.listener.getsockname()
+        self.wait_thread = threading.Thread(
+            target=self._wait_for_pid, name="Process Pid Server"
+        )
+        self.wait_thread.daemon = True
+        self.wait_thread.start()
+
+    def _wait_for_pid(self):
+        try:
+            log.info(
+                "Adapter waiting for connection from launcher on {0}:{1}...",
+                self.host,
+                self.port
+            )
+            sock, (l_host, l_port) = self.listener.accept()
+        finally:
+            self.listener.close()
+        log.info("Launcher connection accepted from {0}:{1}.", l_host, l_port)
+
+        data = b''
+        with sock:
+            try:
+                while True:
+                    d = sock.recv(10)
+                    if d == b'':
+                        break
+                    data += d
+            except OSError:
+                pass
+
+        self.pid = 0 if data == b'' else int(data.decode("ascii"))
+        log.info("Debuggee process Id received: ", self.pid)
+
+    def wait_for_pid(self):
+        self.wait_thread.join()
+        return self.pid
```

Comment:
If the worker directly sets the global `pid` and notifies everybody via `_got_pid` event, this won't be needed, since it's exactly what the caller of this method does.